http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BackgroundProcess.scala

```diff
@@ -0,0 +1,92 @@
+package org.broadinstitute.dsde.workbench.leonardo.monitor
+
+import cats.Show
+import cats.effect.Async
+import cats.syntax.all._
+import fs2.Stream
+import org.broadinstitute.dsde.workbench.model.TraceId
+import org.broadinstitute.dsde.workbench.openTelemetry.OpenTelemetryMetrics
+import org.typelevel.log4cats.StructuredLogger
+
+import java.time.Instant
+import java.util.UUID
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * Defines a common interface that certain background processes should implement.
+ * This type of background processes do the following periodically:
+ * 1. Query the database for a list of candidates
+ * 2. Filter the candidates based on some criteria
+ * 3. Perform some action on the filtered candidates
+ */
+trait BackgroundProcess[F[_], A] {
+  def monitorType: String
+  def interval: FiniteDuration
+
+  /**
+   * Given a candidate, determine whether it should be filtered out.
+   * true: keep the candidate
+   * false: filter out the candidate
+   */
+  def filterCriteria(a: A, now: Instant)(implicit
```

Comment:
Could we combine `filterCriteria` and `dbSource` into a single method called, maybe, `getCandidates`? The caller could handle the DB stuff; I could see there being a wide variety of monitors that might be based on multiple database lookups or querying of other kinds of resources which are not in the db, and it might be simpler to reason about if there were just three stages (get candidates, send logs and metrics, act on candidates).

Reply:
sure